### ISSUE TITLE
1.7.4 segfault in t1ha2_update

### ISSUE THREAD

**Issue by Hello71** [vstakhov/rspamd#2222]
### Classification (Please choose *one* option):

* [x] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [ ] Serious bug
* [ ] Other bug
* [ ] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:
1.7.4

### Operation system, CPU, memory and environment:
Arch Linux, rspamd compiled from AUR

### Debugging information (see details [here](https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed)):
```
#0  0x00005555556f5c55 in t1ha2_update (ctx=ctx@entry=0x555556719670, data=<optimized out>, data@entry=0x555556719998, length=78) at contrib/t1ha/t1ha2.c:273
        need_copy4align = false
        __PRETTY_FUNCTION__ = "t1ha2_update"
#1  0x00005555555c08d6 in rspamd_cryptobox_fast_hash_update (st=st@entry=0x555556719670, data=data@entry=0x555556719998, len=<optimized out>) at src/libcryptobox/cryptobox.c:1496
        rst = 0x555556719670
#2  0x00005555555eb4a3 in rspamd_map_helper_insert_radix_resolve (st=0x555556719650, key=key@entry=0x555555b0cf80, value=value@entry=0x55555576b35c) at src/libutil/map_helpers.c:479
        r = 0x555556719650
        val = 0x555556719980
        vlen = <optimized out>
        k = 1
        nk = 0x555556719998
        res = 1
#3  0x000055555562c87a in rspamd_config_radix_from_ucl (cfg=cfg@entry=0x555555a54250, obj=0x555555b66a50, description=description@entry=0x55555576be21 "Local addresses", target=0x555555a25a70 <local_addrs>, err=err@entry=0x0)
    at src/libserver/cfg_utils.c:1804
        type = <optimized out>
        it = <optimized out>
        cur = <optimized out>
        cur_elt = 0x555555b66a50
        str = 0x555555b0cf80 "192.168.0.0/16, 10.0.0.0/8, 172.16.0.0/12, fd00::/8, 169.254.0.0/16, fe80::/10"
#4  0x00005555555fc6cc in rspamd_config_libs (ctx=0x555555a59790, cfg=cfg@entry=0x555555a54250) at src/libutil/util.c:2135
        secure_ciphers = "HIGH:!aNULL:!kRSA:!PSK:!SRP:!MD5:!RC4"
        r = <optimized out>
        __func__ = "rspamd_config_libs"
#5  0x000055555562d522 in rspamd_config_post_load (cfg=0x555555a54250, opts=<optimized out>) at src/libserver/cfg_utils.c:787
        ts = {tv_sec = 0, tv_nsec = 1}
        ret = 1
        __func__ = "rspamd_config_post_load"
#6  0x00005555555b7ca8 in load_rspamd_config (rspamd_main=0x555555a51750, cfg=0x555555a54250, init_modules=1, opts=(RSPAMD_CONFIG_INIT_URL | RSPAMD_CONFIG_INIT_LIBS | RSPAMD_CONFIG_INIT_SYMCACHE | RSPAMD_CONFIG_INIT_VALIDATE), reload=0)
    at src/rspamd.c:904
        opts = (RSPAMD_CONFIG_INIT_URL | RSPAMD_CONFIG_INIT_LIBS | RSPAMD_CONFIG_INIT_SYMCACHE | RSPAMD_CONFIG_INIT_VALIDATE)
        init_modules = 1
        rspamd_main = 0x555555a51750
        reload = 0
        cfg = 0x555555a54250
```
### Configuration:

I guess it is invalid.

**Comment by vstakhov** [comment]
What platform do you use? Please fill the template properly.

**Comment by Hello71** [comment]
I said, I am using Arch Linux.

**Comment by vstakhov** [comment]
Operation system, **CPU**, memory and environment:

**Comment by leo-yuriev** [comment]
@Hello71, could you provide more info?

Unfortunately,  we cannot reproduce the problem. Moreover, analysis don't shown us the reasonable cause for it. In the other words, this could be a compiler issue or even weirder reason.

So, to help you we needs at least:
 1. CPU registers information, CPU registers information, i.e. output from gdb `info registers` command at the crash.
 2. Actual code of t1ha2_update() function in the your environment, i.e. output from gdb `disassemble t1ha2_update` command at the crash.

Alternatively, a docker image that reproduces the problem, or ssh access, will be appropriate.

**Comment by Hello71** [comment]
```
(gdb) bt full
#0  0x00005555556f5fef in t1ha2_update (ctx=ctx@entry=0x555556708370, data=<optimized out>, data@entry=0x555556708698, length=78) at contrib/t1ha/t1ha2.c:94
        need_copy4align = false
        __PRETTY_FUNCTION__ = "t1ha2_update"
#1  0x00005555555c0586 in rspamd_cryptobox_fast_hash_update (st=st@entry=0x555556708370, data=data@entry=0x555556708698, len=<optimized out>) at src/libcryptobox/cryptobox.c:1496
        rst = 0x555556708370
#2  0x00005555555eaa6a in rspamd_map_helper_insert_radix_resolve (st=0x555556708350, key=key@entry=0x555555afdfa0, value=value@entry=0x55555575b9fc) at src/libutil/map_helpers.c:479
        r = 0x555556708350
        val = 0x555556708680
        vlen = <optimized out>
        k = 3
        nk = 0x555556708698
        res = 1
#3  0x000055555562bd68 in rspamd_config_radix_from_ucl (cfg=cfg@entry=0x555555a45250, obj=0x555555b57a30, description=description@entry=0x55555575c4c1 "Local addresses", target=0x555555a16a70 <local_addrs>, err=err@entry=0x0)
    at src/libserver/cfg_utils.c:1804
        type = <optimized out>
        it = <optimized out>
        cur = <optimized out>
        cur_elt = 0x555555b57a30
        str = 0x555555afdfa0 "192.168.0.0/16, 10.0.0.0/8, 172.16.0.0/12, fd00::/8, 169.254.0.0/16, fe80::/10"
#4  0x00005555555fb81c in rspamd_config_libs (ctx=0x555555a4a790, cfg=cfg@entry=0x555555a45250) at src/libutil/util.c:2135
        secure_ciphers = "HIGH:!aNULL:!kRSA:!PSK:!SRP:!MD5:!RC4"
        r = <optimized out>
        __func__ = "rspamd_config_libs"
#5  0x000055555562ca09 in rspamd_config_post_load (cfg=0x555555a45250, opts=<optimized out>) at src/libserver/cfg_utils.c:787
        ts = {tv_sec = 0, tv_nsec = 1}
        ret = 1
        __func__ = "rspamd_config_post_load"
#6  0x00005555555b79a8 in load_rspamd_config (rspamd_main=0x555555a42750, cfg=0x555555a45250, init_modules=1, opts=(RSPAMD_CONFIG_INIT_URL | RSPAMD_CONFIG_INIT_LIBS | RSPAMD_CONFIG_INIT_SYMCACHE | RSPAMD_CONFIG_INIT_VALIDATE), reload=0)
    at src/rspamd.c:904
        opts = (RSPAMD_CONFIG_INIT_URL | RSPAMD_CONFIG_INIT_LIBS | RSPAMD_CONFIG_INIT_SYMCACHE | RSPAMD_CONFIG_INIT_VALIDATE)
        init_modules = 1
        rspamd_main = 0x555555a42750
        __func__ = "load_rspamd_config"
        reload = 0
        cfg = 0x555555a45250
        __func__ = "load_rspamd_config"
#7  0x00005555555aa02b in main (argc=<optimized out>, argv=<optimized out>, env=<optimized out>) at src/rspamd.c:1300
        i = <optimized out>
        res = 0
        signals = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {0 <repeats 16 times>}}, sa_flags = 0, sa_restorer = 0x0}
        sigpipe_act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {140737348873995, 140737307133880, 93824997402416, 93824997402384, 1436456816, 1, 93824997399696, 93824997402384, 93824997037936, 1, 
              140737488347848, 140737488347864, 93824992822028, 140737488347864, 28, 29}}, sa_flags = 1432005537, sa_restorer = 0x1d}
        pworker = <optimized out>
        type = 1
        control_addr = 0x0
        ev_base = <optimized out>
        term_ev = {ev_evcallback = {evcb_active_next = {tqe_next = 0x0, tqe_prev = 0x0}, evcb_flags = 0, evcb_pri = 0 '\000', evcb_closure = 0 '\000', evcb_cb_union = {evcb_callback = 0x0, evcb_selfcb = 0x0, evcb_evfinalize = 0x0, 
              evcb_cbfinalize = 0x0}, evcb_arg = 0x0}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0x0, tqe_prev = 0x0}, min_heap_idx = 0}, ev_fd = 0, ev_base = 0x0, ev_ = {ev_io = {ev_io_next = {le_next = 0x0, 
                le_prev = 0x0}, ev_timeout = {tv_sec = 0, tv_usec = 0}}, ev_signal = {ev_signal_next = {le_next = 0x0, le_prev = 0x0}, ev_ncalls = 0, ev_pncalls = 0x0}}, ev_events = 0, ev_res = 0, ev_timeout = {tv_sec = 0, tv_usec = 0}}
        int_ev = {ev_evcallback = {evcb_active_next = {tqe_next = 0x0, tqe_prev = 0x0}, evcb_flags = 0, evcb_pri = 0 '\000', evcb_closure = 0 '\000', evcb_cb_union = {evcb_callback = 0x0, evcb_selfcb = 0x0, evcb_evfinalize = 0x0, 
              evcb_cbfinalize = 0x0}, evcb_arg = 0x0}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0x0, tqe_prev = 0x0}, min_heap_idx = 0}, ev_fd = 0, ev_base = 0x0, ev_ = {ev_io = {ev_io_next = {le_next = 0x0, 
                le_prev = 0x0}, ev_timeout = {tv_sec = 0, tv_usec = 0}}, ev_signal = {ev_signal_next = {le_next = 0x0, le_prev = 0x0}, ev_ncalls = 0, ev_pncalls = 0x0}}, ev_events = 0, ev_res = 0, ev_timeout = {tv_sec = 0, tv_usec = 0}}
        cld_ev = {ev_evcallback = {evcb_active_next = {tqe_next = 0x0, tqe_prev = 0x0}, evcb_flags = 0, evcb_pri = 0 '\000', evcb_closure = 0 '\000', evcb_cb_union = {evcb_callback = 0x0, evcb_selfcb = 0x0, evcb_evfinalize = 0x0, 
              evcb_cbfinalize = 0x0}, evcb_arg = 0x0}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0x0, tqe_prev = 0x0}, min_heap_idx = 0}, ev_fd = 0, ev_base = 0x0, ev_ = {ev_io = {ev_io_next = {le_next = 0x0, 
                le_prev = 0x0}, ev_timeout = {tv_sec = 0, tv_usec = 0}}, ev_signal = {ev_signal_next = {le_next = 0x0, le_prev = 0x0}, ev_ncalls = 0, ev_pncalls = 0x0}}, ev_events = 0, ev_res = 0, ev_timeout = {tv_sec = 0, tv_usec = 0}}
        hup_ev = {ev_evcallback = {evcb_active_next = {tqe_next = 0x0, tqe_prev = 0x0}, evcb_flags = 0, evcb_pri = 0 '\000', evcb_closure = 0 '\000', evcb_cb_union = {evcb_callback = 0x0, evcb_selfcb = 0x0, evcb_evfinalize = 0x0, 
              evcb_cbfinalize = 0x0}, evcb_arg = 0x0}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0x0, tqe_prev = 0x0}, min_heap_idx = 0}, ev_fd = 0, ev_base = 0x0, ev_ = {ev_io = {ev_io_next = {le_next = 0x0, 
                le_prev = 0x0}, ev_timeout = {tv_sec = 0, tv_usec = 0}}, ev_signal = {ev_signal_next = {le_next = 0x0, le_prev = 0x0}, ev_ncalls = 0, ev_pncalls = 0x0}}, ev_events = 0, ev_res = 0, ev_timeout = {
            tv_sec = 3704897510905061617, tv_usec = -7444925185664165376}}
        usr1_ev = {ev_evcallback = {evcb_active_next = {tqe_next = 0x7fffffffdfdc, tqe_prev = 0x5555556f1904 <ottery_st_reseed+276>}, evcb_flags = 0, evcb_pri = 0 '\000', evcb_closure = 0 '\000', evcb_cb_union = {evcb_callback = 0x0, 
              evcb_selfcb = 0x0, evcb_evfinalize = 0x0, evcb_cbfinalize = 0x0}, evcb_arg = 0x80}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0x90, tqe_prev = 0x7}, min_heap_idx = 144}, ev_fd = 0, ev_base = 0x9, ev_ = {
            ev_io = {ev_io_next = {le_next = 0x555555a2a048, le_prev = 0x7f}, ev_timeout = {tv_sec = 176, tv_usec = -112}}, ev_signal = {ev_signal_next = {le_next = 0x555555a2a048, le_prev = 0x7f}, ev_ncalls = 176, 
              ev_pncalls = 0xffffffffffffff90}}, ev_events = 2, ev_res = 0, ev_timeout = {tv_sec = 214748364809, tv_usec = 0}}
        control_ev = {ev_evcallback = {evcb_active_next = {tqe_next = 0x0, tqe_prev = 0x6e0000005b}, evcb_flags = 0, evcb_pri = 0 '\000', evcb_closure = 0 '\000', evcb_cb_union = {evcb_callback = 0x770000007c, evcb_selfcb = 0x770000007c, 
              evcb_evfinalize = 0x770000007c, evcb_cbfinalize = 0x770000007c}, evcb_arg = 0x0}, ev_timeout_pos = {ev_next_with_common_timeout = {tqe_next = 0x7ffff5327aa0 <main_arena>, tqe_prev = 0x80}, min_heap_idx = -181241184}, 
          ev_fd = 36432, ev_base = 0x7ffff532c7b8 <__libc_multiple_threads>, ev_ = {ev_io = {ev_io_next = {le_next = 0x555555a421b0, le_prev = 0x9240b79a}, ev_timeout = {tv_sec = -7444925185664165376, tv_usec = 128}}, ev_signal = {
              ev_signal_next = {le_next = 0x555555a421b0, le_prev = 0x9240b79a}, ev_ncalls = -9728, ev_pncalls = 0x80}}, ev_events = -13216, ev_res = 21923, ev_timeout = {tv_sec = 4, tv_usec = 5728953533652119979}}
        term_tv = {tv_sec = 0, tv_usec = 0}
        rspamd_main = <optimized out>
        skip_pid = 0
        __func__ = "main"
(gdb) info reg
rax            0x7f8fa9e8a0cfdb5	574484511317622197
rbx            0x0	0
rcx            0xa908428262f2e159	-6266685753712778919
rdx            0x759eb3dd3c78923c	8475409311541531196
rsi            0xf19597e82ca64a35	-1038757115601728971
rdi            0x615d605c6463685f	7015869744450594911
rbp            0x555555a42750	0x555555a42750
rsp            0x7fffffffdd20	0x7fffffffdd20
r8             0x2c63330d29f3a3a1	3198456291994608545
r9             0x12717080fa3e5c21	1328967064307522593
r10            0x5a04e17619ac9c25	6486557260673489957
r11            0x9c7f9c219c09895d	-7169840413584815779
r12            0x1	1
r13            0x7fffffffe2c0	140737488347840
r14            0x555555a6aec0	93824997568192
r15            0x0	0
rip            0x5555555aa02b	0x5555555aa02b <main+1627>
eflags         0x10202	[ IF RF ]
cs             0x33	51
ss             0x2b	43
ds             0x0	0
es             0x0	0
fs             0x0	0
gs             0x0	0
(gdb) disas t1ha2_update
Dump of assembler code for function t1ha2_update:
   0x00005555556f5ef0 <+0>:	push   %rbp
   0x00005555556f5ef1 <+1>:	mov    %rsp,%rbp
   0x00005555556f5ef4 <+4>:	push   %r15
   0x00005555556f5ef6 <+6>:	push   %r14
   0x00005555556f5ef8 <+8>:	push   %r13
   0x00005555556f5efa <+10>:	mov    %rdx,%r13
   0x00005555556f5efd <+13>:	push   %r12
   0x00005555556f5eff <+15>:	mov    %rdi,%r12
   0x00005555556f5f02 <+18>:	push   %rbx
   0x00005555556f5f03 <+19>:	mov    %rsi,%rbx
   0x00005555556f5f06 <+22>:	and    $0xffffffffffffffe0,%rsp
   0x00005555556f5f0a <+26>:	sub    $0x20,%rsp
   0x00005555556f5f0e <+30>:	mov    0x40(%rdi),%rax
   0x00005555556f5f12 <+34>:	add    %rdx,0x48(%rdi)
   0x00005555556f5f16 <+38>:	test   %rax,%rax
   0x00005555556f5f19 <+41>:	jne    0x5555556f6010 <t1ha2_update+288>
   0x00005555556f5f1f <+47>:	cmp    $0x1f,%r13
   0x00005555556f5f23 <+51>:	jbe    0x5555556f5ff8 <t1ha2_update+264>
   0x00005555556f5f29 <+57>:	mov    %r12,0x18(%rsp)
   0x00005555556f5f2e <+62>:	lea    -0x1f(%rbx,%r13,1),%rax
   0x00005555556f5f33 <+67>:	mov    0x18(%r12),%r11
   0x00005555556f5f38 <+72>:	movabs $0xcb5af53ae3aaac31,%r15
   0x00005555556f5f42 <+82>:	mov    0x10(%r12),%r9
   0x00005555556f5f47 <+87>:	mov    0x8(%r12),%r10
   0x00005555556f5f4c <+92>:	mov    %r13,0x10(%rsp)
   0x00005555556f5f51 <+97>:	movabs $0xc060724a8424f345,%r14
   0x00005555556f5f5b <+107>:	mov    (%r12),%r8
   0x00005555556f5f5f <+111>:	mov    %rax,%r13
   0x00005555556f5f62 <+114>:	nopw   0x0(%rax,%rax,1)
   0x00005555556f5f68 <+120>:	mov    0x8(%rbx),%rax
   0x00005555556f5f6c <+124>:	mov    0x10(%rbx),%r12
   0x00005555556f5f70 <+128>:	add    $0x20,%rbx
   0x00005555556f5f74 <+132>:	mov    -0x8(%rbx),%rdi
   0x00005555556f5f78 <+136>:	lea    (%r11,%r12,1),%rdx
   0x00005555556f5f7c <+140>:	rorx   $0x26,%rax,%rsi
   0x00005555556f5f82 <+146>:	add    %r12,%rax
   0x00005555556f5f85 <+149>:	add    %r10,%rsi
   0x00005555556f5f88 <+152>:	lea    (%r9,%rdi,1),%rcx
   0x00005555556f5f8c <+156>:	add    -0x20(%rbx),%rdi
   0x00005555556f5f90 <+160>:	rorx   $0x38,%rdx,%rdx
   0x00005555556f5f96 <+166>:	xor    %rsi,%r11
   0x00005555556f5f99 <+169>:	rorx   $0x13,%rcx,%rcx
   0x00005555556f5f9f <+175>:	add    %rdi,%rdx
   0x00005555556f5fa2 <+178>:	add    %rcx,%rax
   0x00005555556f5fa5 <+181>:	rorx   $0x39,-0x20(%rbx),%rsi
   0x00005555556f5fac <+188>:	imul   %r15,%rax
   0x00005555556f5fb0 <+192>:	add    %r8,%rsi
   0x00005555556f5fb3 <+195>:	imul   %r14,%rdx
   0x00005555556f5fb7 <+199>:	xor    %rsi,%r9
   0x00005555556f5fba <+202>:	xor    %rax,%r10
   0x00005555556f5fbd <+205>:	xor    %rdx,%r8
   0x00005555556f5fc0 <+208>:	cmp    %rbx,%r13
   0x00005555556f5fc3 <+211>:	ja     0x5555556f5f68 <t1ha2_update+120>
   0x00005555556f5fc5 <+213>:	mov    0x10(%rsp),%r13
   0x00005555556f5fca <+218>:	vmovq  %r9,%xmm2
   0x00005555556f5fcf <+223>:	vmovq  %r8,%xmm3
   0x00005555556f5fd4 <+228>:	mov    0x18(%rsp),%r12
   0x00005555556f5fd9 <+233>:	vpinsrq $0x1,%r11,%xmm2,%xmm1
   0x00005555556f5fdf <+239>:	vpinsrq $0x1,%r10,%xmm3,%xmm0
   0x00005555556f5fe5 <+245>:	vinserti128 $0x1,%xmm1,%ymm0,%ymm0
   0x00005555556f5feb <+251>:	and    $0x1f,%r13d
   0x00005555556f5fef <+255>:	vmovdqa %ymm0,(%r12)
   0x00005555556f5ff5 <+261>:	vzeroupper 
   0x00005555556f5ff8 <+264>:	test   %r13,%r13
   0x00005555556f5ffb <+267>:	jne    0x5555556f6070 <t1ha2_update+384>
   0x00005555556f5ffd <+269>:	lea    -0x28(%rbp),%rsp
   0x00005555556f6001 <+273>:	pop    %rbx
   0x00005555556f6002 <+274>:	pop    %r12
   0x00005555556f6004 <+276>:	pop    %r13
   0x00005555556f6006 <+278>:	pop    %r14
   0x00005555556f6008 <+280>:	pop    %r15
   0x00005555556f600a <+282>:	pop    %rbp
   0x00005555556f600b <+283>:	retq   
   0x00005555556f600c <+284>:	nopl   0x0(%rax)
   0x00005555556f6010 <+288>:	mov    $0x20,%r14d
   0x00005555556f6016 <+294>:	lea    0x20(%rdi,%rax,1),%rdi
   0x00005555556f601b <+299>:	sub    %rax,%r14
   0x00005555556f601e <+302>:	cmp    %r14,%rdx
   0x00005555556f6021 <+305>:	mov    %r14,%r15
   0x00005555556f6024 <+308>:	cmovbe %rdx,%r15
   0x00005555556f6028 <+312>:	mov    %r15,%rdx
   0x00005555556f602b <+315>:	callq  *0x317bef(%rip)        # 0x555555a0dc20
   0x00005555556f6031 <+321>:	mov    0x40(%r12),%rax
   0x00005555556f6036 <+326>:	add    %r15,%rax
   0x00005555556f6039 <+329>:	mov    %rax,0x40(%r12)
   0x00005555556f603e <+334>:	cmp    $0x1f,%rax
   0x00005555556f6042 <+338>:	ja     0x5555556f60c0 <t1ha2_update+464>
   0x00005555556f6044 <+340>:	cmp    %r14,%r13
   0x00005555556f6047 <+343>:	jbe    0x5555556f5ffd <t1ha2_update+269>
   0x00005555556f6049 <+345>:	lea    0xb5e08(%rip),%rcx        # 0x5555557abe58 <__PRETTY_FUNCTION__.12289>
   0x00005555556f6050 <+352>:	mov    $0x102,%edx
   0x00005555556f6055 <+357>:	lea    0xb5e09(%rip),%rsi        # 0x5555557abe65
   0x00005555556f605c <+364>:	lea    0xb5e17(%rip),%rdi        # 0x5555557abe7a
   0x00005555556f6063 <+371>:	callq  *0x317d3f(%rip)        # 0x555555a0dda8
   0x00005555556f6069 <+377>:	nopl   0x0(%rax)
   0x00005555556f6070 <+384>:	mov    %r13,0x40(%r12)
   0x00005555556f6075 <+389>:	lea    0x20(%r12),%rsi
   0x00005555556f607a <+394>:	mov    %r13d,%eax
   0x00005555556f607d <+397>:	cmp    $0x8,%r13d
   0x00005555556f6081 <+401>:	jb     0x5555556f6170 <t1ha2_update+640>
   0x00005555556f6087 <+407>:	mov    %r13d,%eax
   0x00005555556f608a <+410>:	mov    -0x8(%rbx,%rax,1),%rdx
   0x00005555556f608f <+415>:	mov    %rdx,-0x8(%rsi,%rax,1)
   0x00005555556f6094 <+420>:	lea    -0x1(%r13),%eax
   0x00005555556f6098 <+424>:	cmp    $0x8,%eax
   0x00005555556f609b <+427>:	jb     0x5555556f5ffd <t1ha2_update+269>
   0x00005555556f60a1 <+433>:	and    $0xfffffff8,%eax
   0x00005555556f60a4 <+436>:	mov    %eax,%edx
   0x00005555556f60a6 <+438>:	xor    %eax,%eax
   0x00005555556f60a8 <+440>:	mov    %eax,%ecx
   0x00005555556f60aa <+442>:	add    $0x8,%eax
   0x00005555556f60ad <+445>:	mov    (%rbx,%rcx,1),%rdi
   0x00005555556f60b1 <+449>:	mov    %rdi,(%rsi,%rcx,1)
   0x00005555556f60b5 <+453>:	cmp    %edx,%eax
   0x00005555556f60b7 <+455>:	jb     0x5555556f60a8 <t1ha2_update+440>
   0x00005555556f60b9 <+457>:	jmpq   0x5555556f5ffd <t1ha2_update+269>
   0x00005555556f60be <+462>:	xchg   %ax,%ax
   0x00005555556f60c0 <+464>:	mov    0x8(%r12),%r10
   0x00005555556f60c5 <+469>:	mov    0x28(%r12),%r11
   0x00005555556f60ca <+474>:	add    %r15,%rbx
   0x00005555556f60cd <+477>:	sub    %r15,%r13
   0x00005555556f60d0 <+480>:	mov    0x18(%r12),%rdx
   0x00005555556f60d5 <+485>:	mov    (%r12),%rdi
   0x00005555556f60d9 <+489>:	movq   $0x0,0x40(%r12)
   0x00005555556f60e2 <+498>:	mov    0x20(%r12),%r8
   0x00005555556f60e7 <+503>:	rorx   $0x26,%r11,%rsi
   0x00005555556f60ed <+509>:	add    %r10,%rsi
   0x00005555556f60f0 <+512>:	mov    0x10(%r12),%rcx
   0x00005555556f60f5 <+517>:	xor    %rdx,%rsi
   0x00005555556f60f8 <+520>:	mov    0x30(%r12),%rax
   0x00005555556f60fd <+525>:	mov    0x38(%r12),%r9
   0x00005555556f6102 <+530>:	mov    %rsi,0x18(%r12)
   0x00005555556f6107 <+535>:	rorx   $0x39,%r8,%rsi
   0x00005555556f610d <+541>:	add    %rdi,%rsi
   0x00005555556f6110 <+544>:	xor    %rcx,%rsi
   0x00005555556f6113 <+547>:	add    %rax,%rdx
   0x00005555556f6116 <+550>:	mov    %rsi,0x10(%r12)
   0x00005555556f611b <+555>:	lea    (%r9,%rcx,1),%rsi
   0x00005555556f611f <+559>:	lea    (%r11,%rax,1),%rcx
   0x00005555556f6123 <+563>:	rorx   $0x38,%rdx,%rdx
   0x00005555556f6129 <+569>:	lea    (%r8,%r9,1),%rax
   0x00005555556f612d <+573>:	rorx   $0x13,%rsi,%rsi
   0x00005555556f6133 <+579>:	add    %rsi,%rcx
   0x00005555556f6136 <+582>:	movabs $0xcb5af53ae3aaac31,%rsi
   0x00005555556f6140 <+592>:	add    %rdx,%rax
   0x00005555556f6143 <+595>:	imul   %rsi,%rcx
   0x00005555556f6147 <+599>:	movabs $0xc060724a8424f345,%rdx
   0x00005555556f6151 <+609>:	imul   %rdx,%rax
   0x00005555556f6155 <+613>:	xor    %r10,%rcx
   0x00005555556f6158 <+616>:	xor    %rdi,%rax
   0x00005555556f615b <+619>:	mov    %rcx,0x8(%r12)
   0x00005555556f6160 <+624>:	mov    %rax,(%r12)
   0x00005555556f6164 <+628>:	jmpq   0x5555556f5f1f <t1ha2_update+47>
   0x00005555556f6169 <+633>:	nopl   0x0(%rax)
   0x00005555556f6170 <+640>:	and    $0x4,%r13d
   0x00005555556f6174 <+644>:	jne    0x5555556f619c <t1ha2_update+684>
   0x00005555556f6176 <+646>:	test   %eax,%eax
   0x00005555556f6178 <+648>:	je     0x5555556f5ffd <t1ha2_update+269>
   0x00005555556f617e <+654>:	mov    (%rbx),%dl
   0x00005555556f6180 <+656>:	mov    %dl,0x20(%r12)
   0x00005555556f6185 <+661>:	test   $0x2,%al
   0x00005555556f6187 <+663>:	je     0x5555556f5ffd <t1ha2_update+269>
   0x00005555556f618d <+669>:	mov    -0x2(%rbx,%rax,1),%dx
   0x00005555556f6192 <+674>:	mov    %dx,-0x2(%rsi,%rax,1)
   0x00005555556f6197 <+679>:	jmpq   0x5555556f5ffd <t1ha2_update+269>
   0x00005555556f619c <+684>:	mov    (%rbx),%edx
   0x00005555556f619e <+686>:	mov    %edx,0x20(%r12)
   0x00005555556f61a3 <+691>:	mov    -0x4(%rbx,%rax,1),%edx
   0x00005555556f61a7 <+695>:	mov    %edx,-0x4(%rsi,%rax,1)
   0x00005555556f61ab <+699>:	jmpq   0x5555556f5ffd <t1ha2_update+269>
End of assembler dump.
```

I include new `bt full` because of ASLR. I don't know why `=>` doesn't show, perhaps an artifact of gdb logging. You can see the address in the backtrace anyways. I am using gcc 8.1.0 with `-march=native -O2` (and other flags).

**Comment by leo-yuriev** [comment]
@Hello71, I think I know what's going on.
I will try to fix the problem today.
Thank for reporting.

However, I still need to understand why the problem wasn't identified by the tests.

**Comment by Hello71** [comment]
I think it is a gdb bug, bt full moves the selected frame to main for some reason.

https://sourceware.org/bugzilla/show_bug.cgi?id=23203

here is the output without bt full:

```
(gdb) info reg
rax            0x7f8fa9e8a0cfdb5        574484511317622197
rbx            0x5555568c4878   93825012615288
rcx            0xa908428262f2e159       -6266685753712778919
rdx            0x759eb3dd3c78923c       8475409311541531196
rsi            0xf19597e82ca64a35       -1038757115601728971
rdi            0x615d605c6463685f       7015869744450594911
rbp            0x7fffffffdb30   0x7fffffffdb30
rsp            0x7fffffffdae0   0x7fffffffdae0
r8             0x2c63330d29f3a3a1       3198456291994608545
r9             0x12717080fa3e5c21       1328967064307522593
r10            0x5a04e17619ac9c25       6486557260673489957
r11            0x9c7f9c219c09895d       -7169840413584815779
r12            0x5555568c4510   93825012614416
r13            0xe      14
r14            0xc060724a8424f345       -4584538756293004475
r15            0xcb5af53ae3aaac31       -3793450102829700047
rip            0x5555556f5fef   0x5555556f5fef <t1ha2_update+255>
eflags         0x10202  [ IF RF ]
cs             0x33     51
ss             0x2b     43
ds             0x0      0
es             0x0      0
fs             0x0      0
gs             0x0      0
```

**Comment by leo-yuriev** [comment]
Oops, comrades. I suddenly "elephant" not observed. It's obvious.

The compiler used the `vmovdqa ymm, [addr]` (not the `vmovdqa xmm, [addr]` or `movdqa`) instruction, which requires 32-byte alignment of data. However, he is absolutely right, as such alignment is requested while [declaration of the `t1ha_state256_t`](https://github.com/leo-yuriev/t1ha/blob/1e058f9521c25360a0b64582391e3fa4925093aa/t1ha.h#L272-L279) type.

But on the other hand, we can see `ctx=ctx@entry=0x555556719670`, i.e. it is aligned to 16, but not to 32 bytes.

In other words, the bug is in the rspamd's code:
  - internal allocator violate the ABI's alignment requirement;
  - and/or compiler lost information about alignment attributes due some typedefs of opacity techniques;

**Comment by vstakhov** [comment]
Well, in that particular place, it cannot be related to internal allocator at all:

~~~C
struct rspamd_radix_map_helper {
	rspamd_mempool_t *pool;
	khash_t(rspamd_map_hash) *htb;
	radix_compressed_t *trie;
	rspamd_cryptobox_fast_hash_state_t hst;
};
~~~

The structure itself is, indeed, allocated by the internal allocator. However, `hst` field is placed inside and should be properly aligned:

~~~
(lldb) p sizeof(struct rspamd_radix_map_helper)
(unsigned long) $0 = 128
(lldb) p &((struct rspamd_radix_map_helper *)(0))->hst
(rspamd_cryptobox_fast_hash_state_t *) $1 = 0x0000000000000020
~~~

I will check that later.

**Comment by vstakhov** [comment]
Well, I have changed maximum alignment required to 16 bytes.

**Comment by leo-yuriev** [comment]
@Hello71, could you check and confirm/close this issue?

**Comment by Hello71** [comment]
seems to be fixed with 1.7.5. strange, that doesn't include 6f80cd3, but whatever, good enough for me

**Comment by Hello71** [comment]
er, that was because I didn't wait long enough for it to fail. seems to be actually fixed on git master though.
